src/lib/x509/asn1_alt_name.cpp

```diff
@@ -65,170 +76,107 @@ void AlternativeName::add_othername(const OID& oid, std::string_view value, ASN1
-   for(auto i = range.first; i != range.second; ++i) {
-      results.push_back(i->second);
+   for(const auto& othername : this->other_names()) {
+      names.emplace(othername.first.to_formatted_string(), othername.second.value());
    }
-   return results;
+
+   return names;
 }
 
-X509_DN AlternativeName::dn() const {
-   X509_DN dn;
-   auto range = m_alt_info.equal_range("DN");
+std::multimap<std::string, std::string, std::less<>> AlternativeName::get_attributes() const {
+   std::multimap<std::string, std::string, std::less<>> r;
 
-   for(auto i = range.first; i != range.second; ++i) {
-      std::istringstream strm(i->second);
-      strm >> dn;
+   for(const auto& c : this->contents()) {
+      r.emplace(c.first, c.second);
    }
 
-   return dn;
+   return r;
 }
 
-/*
-* Return if this object has anything useful
-*/
-bool AlternativeName::has_items() const {
-   return (!m_alt_info.empty() || !m_othernames.empty());
+bool AlternativeName::has_field(std::string_view attr) const {
+   return !this->get_attribute(attr).empty();
 }
 
-namespace {
+std::string AlternativeName::get_first_attribute(std::string_view type) const {
+   auto attr = this->get_attribute(type);
 
-/*
-* DER encode an AlternativeName entry
-*/
-void encode_entries(DER_Encoder& encoder,
-                    const std::multimap<std::string, std::string, std::less<>>& attr,
-                    std::string_view type,
-                    ASN1_Type tagging) {
-   auto range = attr.equal_range(type);
-
-   for(auto i = range.first; i != range.second; ++i) {
-      if(type == "RFC822" || type == "DNS" || type == "URI") {
-         ASN1_String asn1_string(i->second, ASN1_Type::Ia5String);
-         encoder.add_object(tagging, ASN1_Class::ContextSpecific, asn1_string.value());
-      } else if(type == "IP") {
-         const uint32_t ip = string_to_ipv4(i->second);
-         uint8_t ip_buf[4] = {0};
-         store_be(ip, ip_buf);
-         encoder.add_object(tagging, ASN1_Class::ContextSpecific, ip_buf, 4);
-      } else if(type == "DN") {
-         std::stringstream ss(i->second);
-         X509_DN dn;
-         ss >> dn;
-         encoder.encode(dn);
-      }
+   if(!attr.empty()) {
+      return attr[0];
    }
-}
 
-}  // namespace
-
-/*
-* DER encode an AlternativeName extension
-*/
-void AlternativeName::encode_into(DER_Encoder& der) const {
-   der.start_sequence();
+   return "";
+}
 
-   encode_entries(der, m_alt_info, "RFC822", ASN1_Type(1));
-   encode_entries(der, m_alt_info, "DNS", ASN1_Type(2));
-   encode_entries(der, m_alt_info, "DN", ASN1_Type(4));
-   encode_entries(der, m_alt_info, "URI", ASN1_Type(6));
-   encode_entries(der, m_alt_info, "IP", ASN1_Type(7));
+std::vector<std::string> AlternativeName::get_attribute(std::string_view attr) const {
+   auto set_to_vector = [](const std::set<std::string>& s) {
+      std::vector<std::string> v(s.size());
+      std::copy(s.begin(), s.end(), v.begin());
+      return v;
+   };
```

Comment:
Alternatively:
```cpp
   auto set_to_vector = [](const std::set<std::string>& s) -> std::vector<int> {
      return {s.begin(), s.end()};
   };
```

Reply:
Thanks I would not have expected that to work